machine-runner/src/runner/runner.ts

```diff
@@ -569,6 +700,42 @@ export interface StateOpaque<
   cast(): State<StateName, Payload, Commands>
 }
 
+export namespace StateOpaque {
+  /**
+   * The widest type of StateOpaque. Any other StateOpaque extends this type
+   */
+  export type Any = StateOpaque<string, string, string, any, object>
+
+  /**
+   * Derive StateOpaque type from a SwarmProtocol, a Machine, or a MachineRunner
+   * @example
+   *
+   * const HangarBay = SwarmProtocol.make('HangarBay', ['hangar-bay'], Events.all)
+   * const Door = HangarBay.makeMachine('door')
+   * const Initial = Door.designEmpty().finish()
+   * const machineRunner = createMachineRunner(actyx, where, Passenger.Initial, void 0);
+   *
+   * // Two types below refers to any StateOpaque coming from Door machine, HangarBay protocol
+   * type ThisStateOpaque1 = StateOpaque.Of<typeof machineRunner>;
+   * type ThisStateOpaque2 = StateOpaque.Of<typeof Door>;
+   *
+   * // The type below refers to any StateOpaque coming from HangarBay protocol
+   * type ThisStateOpaque3 = StateOpaque.Of<typeof HangarBay>;
+   */
+  export type Of<
+    M extends
+      | MachineRunner.Any
+      | Machine.Any
+      | SwarmProtocol<string, [any, ...any[]], [any, ...any[]]>,
+  > = M extends MachineRunner<infer S, infer N>
+    ? StateOpaque<S, N>
+    : M extends Machine<infer S, infer N, any>
+    ? StateOpaque<S, N>
+    : M extends SwarmProtocol<infer S, any, any>
+    ? StateOpaque<S, any>
```

Comment:
similar to MachineRunner: where will users need this case?

Reply:
Similar to the MachineRunner one. But I haven't thought about this particular case. 

However, one useful trait that this case might have is that when a user uses a StateOpaque derived from this and tries to call `as` or `is` with the factory that comes from another protocol, the user will catch it.